### ISSUE TITLE
Pyright diagnoses InvalidTypeVarUse where type var is used in autogenerated return type by pyright

### ISSUE THREAD

**Issue by vient** [microsoft/pyright#5228]
Hi, so here is this code
```py
from abc import abstractmethod
from typing import Callable, Concatenate, Iterable, ParamSpec, Protocol, TypeVar


P = ParamSpec('P')
V = TypeVar('V', covariant=True)


class _PaginatedResponse(Protocol[V]):
    @property
    @abstractmethod
    def values(self) -> Iterable[V]:
        raise NotImplementedError


def _paginated_api(f: Callable[Concatenate[int, P], _PaginatedResponse[V]]):
    def paginated_api_decorator(limit: int | None = None, *args: P.args, **kwargs: P.kwargs) -> list[V]:
        # actual implementation omitted 
        return list(f(1, *args, **kwargs).values)

    return paginated_api_decorator

# @_paginated_api
# def f(...): ...
```
I omitted the return type hint for `_paginated_api` function because it is trivial from its source and cumbersome. Pyright correctly deduces return types for functions decorated with `_paginated_api` but generates a warning on line with `def _paginated_api...`:
```
TypeVar "V" appears only once in generic function signature Pylance(reportInvalidTypeVarUse)
(constant) V: Type[V@_paginated_api]
```
What's interesting, it does not produce this warning for the next line, where `V` appears only once in return type hint.

Deduced `_paginated_api` signature is
```py
def _paginated_api(f: (int, **P@_paginated_api) -> _PaginatedResponse[V@_paginated_api]) -> ((limit: (int | None) = None, *args: P.args, **kwargs: P.kwargs) -> list[V@_paginated_api])
```
so `V` is used more than once.

Is the warning issued incorrectly? If not, is it possible to instruct pyright to issue warnings considering deduced type hints and not only the ones explicitly written?

**Comment by erictraut** [comment]
I presume you've enabled the `reportInvalidTypeVarUse` diagnostic check. This isn't on by default in "basic" type checking mode. It's a strict-mode check and is fairly opinionated in how it works. It's working as intended in this case. You would need to provide an explicit return type annotation to eliminate the error. Or you can use `# pyright: ignore[reportInvalidTypeVarUse]` if you want to suppress it.

The reason you don't see an error in the second case is that `V` is bound to an outer scope, so it's OK for it to be used by itself in an inner scope.
